app/tests/spec/models/attached-clients.js

```diff
@@ -0,0 +1,223 @@
+  var Constants = require('lib/constants');
+  var Notifier = require('lib/channels/notifier');
+  var p = require('lib/promise');
+  var sinon = require('sinon');
+  var User = require('models/user');
+
+  describe('models/attached-clients', function () {
+    var attachedClients;
+    var notifier;
+    var user;
+
+    beforeEach(function () {
+      notifier = new Notifier();
+      user = new User();
+
+      attachedClients = new AttachedClients([], {
+        notifier: notifier
+      });
+    });
+
+    describe('properly orders the attached clients', function () {
+      var now = Date.now();
+
+      beforeEach(function () {
+        attachedClients.set([
+          {
+            clientType: Constants.CLIENT_TYPE_DEVICE,
+            isCurrentDevice: false,
+            lastAccessTime: null,
+            name: 'xi'
+          },
+          {
+            clientType: Constants.CLIENT_TYPE_DEVICE,
+            isCurrentDevice: false,
+            lastAccessTime: null,
+            name: 'xi'
+          },
+          {
+            clientType: Constants.CLIENT_TYPE_DEVICE,
+            isCurrentDevice: true,
+            lastAccessTime: now - 20,
+            name: 'zeta'
+          },
+          {
+            clientType: Constants.CLIENT_TYPE_OAUTH_APP,
+            isCurrentDevice: false,
+            lastAccessTime: now - 10,
+            name: 'mu'
+          },
+          {
+            clientType: Constants.CLIENT_TYPE_DEVICE,
+            isCurrentDevice: false,
+            lastAccessTime: now,
+            name: 'tau'
+          },
+          {
+            clientType: Constants.CLIENT_TYPE_DEVICE,
+            isCurrentDevice: false,
+            lastAccessTime: now,
+            name: 'sigma'
+          },
+          {
+            clientType: Constants.CLIENT_TYPE_DEVICE,
+            isCurrentDevice: false,
+            lastAccessTime: now - 20,
+            name: 'theta'
+          },
+          {
+            clientType: Constants.CLIENT_TYPE_OAUTH_APP,
+            lastAccessTime: null,
+            name: 'an oauth'
+          }
+        ]);
+      });
+
+      it('places the `current` device first', function () {
+        assert.equal(attachedClients.at(0).get('name'), 'zeta');
+      });
+
+      it('sorts those with lastAccessTime next, by access time (descending)', function () {
+        assert.equal(attachedClients.at(1).get('name'), 'sigma');
+        assert.equal(attachedClients.at(2).get('name'), 'tau');
+        assert.equal(attachedClients.at(3).get('name'), 'mu');
+        assert.equal(attachedClients.at(4).get('name'), 'theta');
+
+      });
+
+      it('sorts the rest alphabetically', function () {
+        assert.equal(attachedClients.at(5).get('name'), 'an oauth');
+        assert.equal(attachedClients.at(6).get('name'), 'xi');
+        assert.equal(attachedClients.at(7).get('name'), 'xi');
+      });
+    });
+
+    describe('device name change', function () {
+      beforeEach(function () {
```

Comment:
It was here before I started adding OAuth apps, i think we can remove this block for now.